### PR TITLE
enhance:  index meta use independent rather than global meta lock

### PR DESCRIPTION
**Pull request by jaime0815**
issue: https://github.com/milvus-io/milvus/issues/30837

### CODE REVIEW

File: internal/datacoord/meta.go [hunk 1]
```diff
@@ -1274,6 +1248,25 @@ func (m *meta) GetEarliestStartPositionOfGrowingSegments(label *CompactionGroupL
 	return earliest
 }
 
+func (m *meta) GetHasUnindexTaskSegments() []*SegmentInfo {
```
Comment (xiaocai2333) [inline internal/datacoord/meta.go:1251]: Would it be better to refactor this into a utility function? The implementation of utility function could be: 
```
func (s *Server)GetHasUnindexTaskSegments() []*SegmentInfo{
    flushedSegments := s.meta.SelectSegment(func(seg *SegmentInfo) bool {
         return isFlush(seg)
    })

    for _, segID := range flushedSegments {
        s.indexMeta.IsUnindexSegment(seg)
        ....
    }
}
```

This way, we can avoid having two layers of locks.

**Review by xiaocai2333 — APPROVED** [review]
LGTM with some easy comments

File: internal/datacoord/server.go [hunk 2]
```diff
@@ -624,7 +624,13 @@ func (s *Server) initMeta(chunkManager storage.ChunkManager) error {
 	reloadEtcdFn := func() error {
 		var err error
 		catalog := datacoord.NewCatalog(s.kv, chunkManager.RootPath(), metaRootPath)
-		s.meta, err = newMeta(s.ctx, catalog, chunkManager)
+
+		// TODO: reload index and segment meta concurrently
+		indexMeta, err := newIndexMeta(s.ctx, catalog)
```
Comment (czs007) [inline internal/datacoord/server.go:629]: move this into newMeta?

File: internal/datacoord/index_meta.go [hunk 3]
```diff
@@ -745,3 +838,15 @@ func (m *meta) GetMetasByNodeID(nodeID UniqueID) []*model.SegmentIndex {
 	}
 	return metas
 }
+
+func (m *segmentIndexMeta) DropSegmentIndex(segmentID UniqueID, indexID UniqueID) {
```
Comment (czs007) [inline internal/datacoord/index_meta.go:842]: delete this code

**Pull request merged**